ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+
+  let submode_mvmv (type a) ?log (dst : a C.obj) (Amorphvar (v, f) as mv)
+      (Amorphvar (u, g) as mu) =
+    if C.le dst (mupper dst mv) (mlower dst mu)
+    then Ok ()
+    else if eq_morphvar dst mv mu
+    then Ok ()
+    else
+      (* we have f v <= g u which gives g' (f v) <= u. On the other hand, we
+         also have v <= f' (g u) *)
+      match submode_mvc ?log dst mv (mupper dst mu) with
+      | Error a -> Error (a, mupper dst mu)
+      | Ok () -> (
+        match submode_cmv ?log dst (mlower dst mv) mu with
+        | Error a -> Error (mlower dst mv, a)
+        | Ok () ->
+          let g' = C.left_adjoint dst g in
+          let src = C.src dst g in
+          let g'f = C.compose src g' (C.disallow_right f) in
+          let x = Amorphvar (v, g'f) in
+          if not (List.exists (fun mv -> eq_morphvar src mv x) u.vlower)
+          then set_vlower ?log u (x :: u.vlower);
+          Ok ())
+
+  let fresh (type a) (obj : a C.obj) =
+    { upper = C.max obj; lower = C.min obj; vlower = [] }
+
+  let newvar obj = Amodevar (Amorphvar (fresh obj, C.id))
+
+  let submode_try (type a r l) ?(logging = true) (obj : a C.obj)
+      (a : (a, allowed * r) mode) (b : (a, l * allowed) mode) =
+    let log = if logging then Some (ref []) else None in
+    let submode_cc left right =
+      if C.le obj left right then Ok () else Error { left; right }
+    in
+    let submode_mvc v right =
+      Result.map_error
+        (fun left ->
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_mvc ?log obj v right)
+    in
+    let submode_cmv left v =
+      Result.map_error
+        (fun right ->
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_cmv ?log obj left v)
+    in
+    let submode_mvmv v u =
+      Result.map_error
+        (fun (left, right) ->
+          assert (not (C.le obj left right));
+          { left; right })
+        (submode_mvmv ?log obj v u)
+    in
+    let r =
+      match a, b with
+      | Amode left, Amode right -> submode_cc left right
+      | Amodevar v, Amode right -> submode_mvc v right
+      | Amode left, Amodevar v -> submode_cmv left v
+      | Amodevar v, Amodevar u -> submode_mvmv v u
+      | Amode a, Amodemeet (b, mvs) ->
+        Result.bind (submode_cc a b) (fun () ->
+            find_error (fun mv -> submode_cmv a mv) mvs)
+      | Amodevar mv, Amodemeet (b, mvs) ->
+        Result.bind (submode_mvc mv b) (fun () ->
+            find_error (fun mv' -> submode_mvmv mv mv') mvs)
+      | Amodejoin (a, mvs), Amode b ->
+        Result.bind (submode_cc a b) (fun () ->
+            find_error (fun mv' -> submode_mvc mv' b) mvs)
+      | Amodejoin (a, mvs), Amodevar mv ->
+        Result.bind (submode_cmv a mv) (fun () ->
+            find_error (fun mv' -> submode_mvmv mv' mv) mvs)
+      | Amodejoin (a, mvs), Amodemeet (b, mus) ->
+        (* TODO: mabye create a intermediate variable? *)
```

Comment:
Not sure what you mean here. Elaborate?

Reply:
This is the case where we say `join [v0,v1,...,v10] <= meet [u0,u1,...,u10]`.
Currently this  translates to 10*10=100 calls of `submode`, which is slow.
I think a smarter way might be to create an intermediate variable, and instead say that:
```
join [v0,v1,...,v10] <= v
v <= meet [u0,u1,...,u10]
```

Reply:
Ah, yes. That does sound like an improvement.

Reply:
On a second thought: secretly creating a new mode variable every time `join <= meet` is called sounds unexpected. Like, people would think the side-effect of `submode` is just the mutation of some bounds, not variable creations. 

I think it might be wise to look at the profile before deciding if this need optimisation. What do you think?

Reply:
I did some `perf-memtrace` profiling, and `submode join meet` accounts for zero sample (among 10K sample) in `ocamlc typecore.ml`. So I assume this is not a big concern.